libraries/Microsoft.Bot.StreamingExtensions/Payloads/Assemblers/ContentStreamAssembler.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+
+namespace Microsoft.Bot.StreamingExtensions.Payloads
+{
+    internal class ContentStreamAssembler : PayloadAssembler
+    {
+        private readonly IStreamManager _streamManager;
+
+        public ContentStreamAssembler(IStreamManager streamManager, Guid id)
+            : base(id)
+        {
+            ContentType = null;
+            ContentLength = null;
+            _streamManager = streamManager;
+        }
+
+        public ContentStreamAssembler(IStreamManager streamManager, Guid id, string type, int? length)
+            : this(streamManager, id)
+        {
+            ContentType = type;
+            ContentLength = length;
+        }
+
+        public int? ContentLength { get; set; }
+
+        public string ContentType { get; set; }
+
+        public override Stream CreatePayloadStream()
+        {
+            return new PayloadStream(this);
+        }
+
+        public override void OnReceive(Header header, Stream stream, int contentLength)
+        {
+            base.OnReceive(header, stream, contentLength);
+
+            if (End)
+            {
+                ((PayloadStream)stream).DoneProducing();
+            }
```

Comment:
Why is the cast needed? More relevaltly, how do you know it'll work?

Reply:
I know it will work because it's casting a class to a class that extends it. The cast is needed thanks to the abstract Assembler base class using Streams in place of any of the four extension classes that the assemblers implementing it deal with. I'll think on ways to make it cleaner.